library/file.lisp

```diff
@@ -0,0 +1,671 @@
+   #:Create
+
+   #:StreamOptions
+   #:Input
+   #:Output
+   #:TwoWay
+   #:Broadcast
+
+   #:Streamable
+   #:open
+   #:close
+   #:read
+   #:read-sequence
+   #:write
+   #:write-sequence
+
+   #:with-open-file
+   #:read-char
+   #:read-byte
+   #:write-char
+   #:write-byte
+   #:write-string
+
+   #:force
+   #:finish
+   #:clear
+   #:file-position
+   #:set-file-position
+   
+   #:read-file-string
+   #:read-file-lines
+   ))
+
+(in-package #:coalton-library/file)
+
+(named-readtables:in-readtable coalton:coalton)
+
+#+coalton-release
+(cl:declaim #.coalton-impl/settings:*coalton-optimize-library*)
+
+;;;
+;;; File Error handling
+;;;
+;;; File errors will return results of type `(Result FileError :a)`
+;;;
+
+
+;;;
+;;; Pathnames
+;;;
+
+(coalton-toplevel
+
+  (repr :native cl:pathname)
+  (define-type Pathname
+    "Pathname object. Uses `cl:pathname`")
+
+  (repr :native cl:condition)
+  (define-type LispCondition
+    "Condition for lisp error handling. Uses `cl:condition`."))
+
+(cl:defmacro handle-file-function ((func cl:&rest args))
+  "A macro for handling potentially erroring lisp file operations. Automatically returns the lisp condition if one is thrown."
+  `(cl:handler-case (Ok (,func ,@args))
+     (cl:error (c) (Err (LispError c)))))
+
+(coalton-toplevel
+
+  (define-type FileError
+    "Errors for file functions."
+    (FileError String)
+    (PathError String Pathname)
+    (LispError LispCondition)
+    EOF)
+
+  (define-instance (Signal FileError)
+    (define (error ferr)
+      (match ferr
+        ((FileError str1)
+         (error (lisp String (str1)
+                  (cl:format cl:nil "File Error:~%~%~a" str1))))
+        ((PathError str path)
+         (error (lisp String (str path)
+                  (cl:format cl:nil "Path Error:~%~%~a ~a" str path))))
+        ((LispError c)
+         (error (lisp String (c)
+                  (cl:format cl:nil "Captured Lisp Error:~%~%~a" c))))
+        ((EOF)
+         (error (lisp String ()
+                  (cl:format cl:nil "Error:~%~%End of File"))))))))
+
+;;;
+;;; Pathname type
+;;;
+
+(coalton-toplevel
+
+  (define-instance (Into String Pathname)
+    (define (into s)
+      (lisp Pathname (s)
+        (cl:pathname s))))
+
+  (define-instance (TryInto Pathname FileError Pathname)
+    (define (tryinto x)
+      (Ok x)))
+  
+  (declare merge ((Into :a Pathname) (Into :b Pathname) => :a -> :b -> (Result FileError Pathname)))
+  (define (merge path1 path2)
+    "Merges two pathnames together. The order is important- for adding subdirectories, make sure to put the subpath (the one without the leading `/`) in `path1`."
+    (do
+     (let p1 = (the Pathname (into path1)))
+     (let p2 = (the Pathname (into path2)))
+      (pure (lisp Pathname (p1 p2)
+              (cl:merge-pathnames p1 p2))))))
+
+;;;
+;;; Handling existential path queries
+;;;
+
+(coalton-toplevel
+
+  (declare valid-directory-pathname? ((Into :a Pathname) => :a -> (Result FileError Boolean)))
```

Comment:
`valid-directory-pathname?` -> `directory-pathname?`

this function doesn't really validate anything